### PR TITLE
Improve extensions performance & add OpenTelemetry

### LINKED ISSUE #649: Replace OpenTracing extension with one for OpenTelemetry?
https://opentracing.io/ now carries a banner announcing that it's been supplanted by OpenTelemetry

Are you interested in a PR that adds OpenTelemetry based tracing?

### PR DESCRIPTION
**Pull request by rafalp**
This PR removes separate `ExtensionSync` and synchronous extensions, making extensions work in both async and sync context based on `next_` being coroutine function or not.

Fixes #649

# TODO

- [x] OpenTracing
- [x] OpenTelemetry
- [x] ApolloTracing (needs test for async and sync context)
- [x] Update benchmarks

### CODE REVIEW

File: ariadne/contrib/tracing/opentelemetry.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+        with self._tracer.start_as_current_span(
+            info.field_name, context=set_span_in_context(self._root_span)
+        ) as span:
+            span.set_attribute("component", "GraphQL")
+            span.set_attribute("graphql.parentType", info.parent_type.name)
+            span.set_attribute("graphql.path", graphql_path)
+
+            if info.operation.name:
+                span.set_attribute("graphql.operation", info.operation.name.value)
```
Comment (dacevedo12) [inline ariadne/contrib/tracing/opentelemetry.py:79]:
```suggestion
                span.set_attribute("graphql.operation.name", info.operation.name.value)
```
per https://opentelemetry.io/docs/specs/otel/trace/semantic_conventions/instrumentation/graphql/

File: ariadne/contrib/tracing/opentracing.py [hunk 2]
```diff
@@ -1,149 +1,126 @@
...
     def request_started(self, context: ContextValue):
-        self._root_scope = self._tracer.start_active_span("GraphQL Query")
-        self._root_scope.span.set_tag(tags.COMPONENT, "graphql")
+        if self._root_span_name:
+            if callable(self._root_span_name):
+                root_span_name = self._root_span_name(context)
+            else:
+                root_span_name = self._root_span_name
+        else:
+            root_span_name = "Operation"
```
[the remark this answers is not in the document]
Comment (dacevedo12) [reply ariadne/contrib/tracing/opentracing.py:41]: should also include `graphql.document` but I'm not sure if that can be referenced at this function

[the remark this answers is not in the document]
Comment (rafalp) [reply ariadne/contrib/tracing/opentracing.py]: I hate that `graphql.document` thing. What if its big or contains personal data embed in query body? Ugh.

**Comment by rafalp** [comment]
Thanks for your feedback, @dacevedo12, I've did some tweaks.

**Review by mat-sop — APPROVED** [review]

**Pull request merged**